### PR TITLE
HUD Mod display

### PR DESCRIPTION
**Pull request by peppy**
- Builds off of (and supersedes) https://github.com/ppy/osu/pull/702.
- Adjusts namespaces of HUD elements to better group them (they are no longer related to Ruleset logic.
- Improves ModIcon/ModSection handling to avoid the need for ModIcons with null mod values.

### CODE REVIEW

**Comment by peppy** [comment]
This should be ready for review 👍

File: osu.Game/Overlays/Mods/ModButtonEmpty.cs [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+using OpenTK;
+using osu.Framework.Graphics.Containers;
+using osu.Game.Rulesets.Mods;
+
+namespace osu.Game.Overlays.Mods
+{
+    /// <summary>
+    /// A mod button used exclusively for providing an empty space the size of a mod button.
+    /// </summary>
+    public class ModButtonEmpty : Container
```
Comment (huoyaoyuan) [inline osu.Game/Overlays/Mods/ModButtonEmpty.cs]: curious as a name of base class

**Comment by huoyaoyuan** [comment]
I dislike the type `MultiMod`.
A `ModButton` can receive multiple mods for select, so it should receive a `Mod[]` directly.
I think it should expose both `Mod` and `Mods` property for set, as many buttons are just receiving one.

File: osu.Game/Overlays/Mods/ModButtonEmpty.cs (hunk 1 above)
Comment (peppy) [inline osu.Game/Overlays/Mods/ModButtonEmpty.cs]: please provide a suggestion. i think this is a step up from using `ModButtonBase` (i'm really against using the base suffix, even though we use it in a few places).

File: osu.Game/Overlays/Mods/ModButton.cs [hunk 2]
```diff
@@ -63,19 +67,8 @@ private int selectedIndex
...
-            {
-                if (value == buttonColour) return;
-                buttonColour = value;
-                foreach (ModIcon icon in iconsContainer.Children)
-                {
-                    icon.Colour = value;
-                }
-            }
+            get { return buttonColour; }
+            set { buttonColour = value; }
```
Comment (smoogipooo) [inline osu.Game/Overlays/Mods/ModButton.cs]: No longer a need for property.

File: osu.Game/Rulesets/UI/ModIcon.cs [hunk 3]
```diff
@@ -74,11 +49,33 @@ public ModIcon()
...
+            switch (mod.Type)
+            {
+                case ModType.DifficultyIncrease:
+                    return OsuColour.FromHex(@"ffcc22");
+                case ModType.DifficultyReduction:
+                    return OsuColour.FromHex(@"88b300");
+                case ModType.Special:
+                    return OsuColour.FromHex(@"66ccff");
+
+                default: return Color4.White;
```
Comment (smoogipooo) [inline osu.Game/Rulesets/UI/ModIcon.cs]: This is a weird structure to this default. Please follow the case statement structure (incl. removing the newline before it).

File: osu.Game/Screens/Play/Player.cs [hunk 4]
```diff
@@ -354,6 +356,10 @@ protected override void OnEntering(Screen last)
 
             hitRendererContainer.Alpha = 0;
             hitRendererContainer.FadeIn(750, EasingTypes.OutQuint);
+
+            if (!HitRenderer.HasReplayLoaded)
+                using (hudOverlay.ModDisplay.BeginDelayedSequence(2000))
+                    hudOverlay.ModDisplay.FadeOut(200);
```
Comment (smoogipooo) [inline osu.Game/Screens/Play/Player.cs]: This is HUD-specific and should not be a part of Player.

File: osu.Game/Rulesets/UI/ModIcon.cs [hunk 5]
```diff
@@ -74,11 +49,33 @@ public ModIcon()
...
+        private Color4 getBackgroundColourFromMod(Mod mod)
+        {
+            switch (mod.Type)
+            {
+                case ModType.DifficultyIncrease:
+                    return OsuColour.FromHex(@"ffcc22");
+                case ModType.DifficultyReduction:
+                    return OsuColour.FromHex(@"88b300");
+                case ModType.Special:
+                    return OsuColour.FromHex(@"66ccff");
```
Comment (smoogipooo) [inline osu.Game/Rulesets/UI/ModIcon.cs]: Use colours from OsuColour.

**Review by smoogipoo — CHANGES_REQUESTED** [review]

File: osu.Game/Overlays/Mods/ModButton.cs (hunk 2 above)
Comment (EVAST9919) [inline osu.Game/Overlays/Mods/ModButton.cs]: it's needable, because we need that colour after switching back from selected state

Comment (smoogipooo) [inline osu.Game/Overlays/Mods/ModButton.cs]: I mean all this property is doing is wrapping a local private.

File: osu.Game/Rulesets/UI/ModIcon.cs [hunk 6]
```diff
@@ -74,11 +49,33 @@ public ModIcon()
...
-                    TextSize = 20
+                    TextSize = 20,
+                    Icon = mod.Icon
                 },
             };
 
             reapplySize();
         }
+
+        private void reapplySize()
```
Comment (smoogipooo) [inline osu.Game/Rulesets/UI/ModIcon.cs]: Function is unneeded. Just set the properties in here directly in the ctor.

File: osu.Game/Screens/Play/Player.cs (hunk 4 above)
Comment (peppy) [inline osu.Game/Screens/Play/Player.cs]: how shall we handle this? pass in a bool `HasReplay` to hudoverlay?

Comment (smoogipooo) [inline osu.Game/Screens/Play/Player.cs]: That's one possibility, although I was leaning more towards using hudOverlay.BindHitRenderer.

Comment (peppy) [inline osu.Game/Screens/Play/Player.cs]: ooh, didn't realise it's already passed in.

**Review by smoogipoo — APPROVED** [review]
👍

**Pull request merged**